pkg/protocol/engine/ledger/ledger/vm.go

```diff
@@ -24,53 +24,82 @@ func (l *Ledger) executeStardustVM(_ context.Context, stateTransition mempool.Tr
+	commitment := tx.CommitmentInput()
+
+	if (len(rewardInputs) > 0 || len(bicInputs) > 0) && commitment == nil {
+		return nil, xerrors.Errorf("commitment input required with Reward or BIC input")
 	}
-	for _, inp := range commitmentInputs {
-		c, err := l.loadCommitment(inp.CommitmentID)
+
+	var loadedCommitment *iotago.Commitment
+	if commitment != nil {
+		loadedCommitment, err = l.loadCommitment(commitment.CommitmentID)
 		if err != nil {
 			return nil, xerrors.Errorf("could not load commitment: %w", err)
 		}
-		commitmentInputSet[inp.CommitmentID] = c
+
+		resolvedInputs.CommitmentInput = loadedCommitment
 	}
 
-	resolvedInputs := iotagovm.ResolvedInputs{
-		InputSet:           inputSet,
-		BICInputSet:        bicInputSet,
-		CommitmentInputSet: commitmentInputSet,
+	bicInputSet := make(iotagovm.BICInputSet)
+	for _, inp := range bicInputs {
+		accountData, exists, accountErr := l.accountsLedger.Account(inp.AccountID, loadedCommitment.Index)
+		if accountErr != nil {
+			return nil, xerrors.Errorf("could not get BIC input for account %s in slot %d: %w", inp.AccountID, loadedCommitment.Index, accountErr)
+		}
+		if !exists {
+			return nil, xerrors.Errorf("BIC input does not exist for account %s in slot %d", inp.AccountID, loadedCommitment.Index)
+		}
+
+		bicInputSet[inp.AccountID] = accountData.Credits.Value
+	}
+	resolvedInputs.BICInputSet = bicInputSet
+
+	rewardInputSet := make(iotagovm.RewardsInputSet)
+	// when refactoring this to be resolved by the mempool, the function that resolves ContextInputs should receive a slice with promises of UTXO inputs and wait until the necessary
+	for _, inp := range rewardInputs {
+		switch castOutput := inputStates[inp.Index].Output().(type) {
+		case *iotago.AccountOutput:
+			stakingFeature := castOutput.FeatureSet().Staking()
+			if stakingFeature == nil {
+				return nil, xerrors.Errorf("cannot claim rewards from an AccountOutput at index %d without staking feature", inp.Index)
+			}
+			reward, rewardErr := l.epochGadget.ValidatorReward(castOutput.AccountID, stakingFeature.StakedAmount, stakingFeature.StartEpoch, stakingFeature.EndEpoch)
+			if rewardErr != nil {
+				return nil, xerrors.Errorf("failed to get Validator reward for AccountOutput at index %d (StakedAmount: %d, StartEpoch: %d, EndEpoch: %d", castOutput.AccountID, stakingFeature.StakedAmount, stakingFeature.StartEpoch, stakingFeature.EndEpoch)
+			}
+
+			rewardInputSet[castOutput.AccountID] += reward
+		case *iotago.DelegationOutput:
+			reward, rewardErr := l.epochGadget.DelegatorReward(castOutput.ValidatorID, castOutput.DelegatedAmount, castOutput.StartEpoch, castOutput.EndEpoch)
+			if rewardErr != nil {
+				return nil, xerrors.Errorf("failed to get Delegator reward for DelegationOutput at index %d (StakedAmount: %d, StartEpoch: %d, EndEpoch: %d", castOutput.DelegationID, castOutput.DelegatedAmount, castOutput.StartEpoch, castOutput.EndEpoch)
+			}
+
+			rewardInputSet[castOutput.DelegationID] += reward
```

Comment:
```suggestion
			delegationEnd := castOutput.EndEpoch

			delegationID := castOutput.DelegationID
			if delegationID.Empty() {
				delegationID = iotago.DelegationIDFromOutputID(inputStates[inp.Index].OutputID())
			}

			if delegationEnd == 0 {
				delegationEnd = l.protocolParameters.TimeProvider().EpochFromSlot(loadedCommitment.Index) - iotago.EpochIndex(1)
			}

			reward, rewardErr := l.epochGadget.DelegatorReward(castOutput.ValidatorID, castOutput.DelegatedAmount, castOutput.StartEpoch, delegationEnd)
			if rewardErr != nil {
				return nil, xerrors.Errorf("failed to get Delegator reward for DelegationOutput at index %d (StakedAmount: %d, StartEpoch: %d, EndEpoch: %d", delegationID, castOutput.DelegatedAmount, castOutput.StartEpoch, castOutput.EndEpoch)
			}

			rewardInputSet[delegationID] = reward
```

- DelegationID may be zero, if so, compute from output ID.
- endEpoch may be zero on the output, if so, use current epoch index - 1.
- Use `=` over `+=` as mentioned above.